airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml

```diff
@@ -118,6 +118,29 @@ definitions:
         field_name: "{{ parameters.get('cursor_field', 'updated') }}"
         inject_into: request_parameter
 
+  base_incremental_checkpoint_stream:
+    $ref: "#/definitions/base_stream"
+    retriever: "#/definitions/base_retriever"
+    incremental_sync:
+      type: CustomIncrementalSync
+      class_name: source_klaviyo.components.datetime_based_cursor.KlaviyoCheckpointDatetimeBasedCursor
+      cursor_field: "{{ parameters.get('cursor_field', 'updated') }}"
+      start_datetime: "{{ config.get('start_date', '2012-01-01T00:00:00Z') }}"
```

Comment:
Is it safer to move the default value to the config to avoid possible date-time type of bugs in the future?

Reply:
Hey @bazarnov, I was playing the default value in spec, but I was surprised that doing something like the below and removing the default date value from the start_datetime's in the incremental_sync setup will throw the an error:

```yaml
      start_date:
        default: '2012-01-01T00:00:00Z'
        type: string
        title: "Start Date"
        ...
``` 

```json
{
  "api_key": "api_key",
}
```

```shell
 poetry run source-klaviyo read --config secrets/config.json --catalog integration_tests/configured_catalog.json
```

```json
{
  "type": "TRACE",
  "trace": {
    "type": "ERROR",
    "emitted_at": 1728905129118.228,
    "error": {
      "message": "Something went wrong in the connector. See the logs for more details.",
      "internal_message": "time data 'None' does not match format '%Y-%m-%dT%H:%M:%S%z'",
      "stack_trace": "Traceback (most recent call last):
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py" (133): read
                yield from self._read_stream(
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py" (239): _read_stream
                for record_data_or_message in record_iterator:
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/streams/core.py" (176): read
                checkpoint_reader = self._get_checkpoint_reader(
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/streams/core.py" (438): _get_checkpoint_reader
                mappings_or_slices = self.stream_slices(
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py" (158): stream_slices
                return self.retriever.stream_slices()
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py" (426): stream_slices
                return self.stream_slicer.stream_slices()
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py" (169): stream_slices
                start_datetime = self._calculate_earliest_possible_value(self.select_best_end_datetime())
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py" (179): _calculate_earliest_possible_value
                earliest_possible_start_datetime = min(self._start_datetime.get_datetime(self.config), end_datetime)
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/min_max_datetime.py" (58): get_datetime
                time = self._parser.parse(str(self.datetime.eval(config, **additional_parameters)), datetime_format)  # type: ignore # datetime is always cast to an interpolated string
        "/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/datetime_parser.py" (35): parse
                parsed_datetime = datetime.datetime.strptime(str(date), format)
        "/Users/aldogonzalez/.pyenv/versions/3.10.14/lib/python3.10/_strptime.py" (568): _strptime_datetime
                tt, fraction, gmtoff_fraction = _strptime(data_string, format)
        "/Users/aldogonzalez/.pyenv/versions/3.10.14/lib/python3.10/_strptime.py" (349): _strptime
                raise ValueError("time data %r does not match format %r" %
        Error  ValueError: time data 'None' does not match format '%Y-%m-%dT%H:%M:%S%z'",
      "failure_type": "system_error",
      "stream_descriptor": {
        "name": "events"
      }
    }
  }
}

``` 

I would need to investigate more. With this setup, I'm almost sure that the UI would show a default value for the start date, but I expected that the value would also be interpreted by the read command at some point.

Reply:
This is because you removed the `start_date` from the `config`, but for local testing it should be present, because `spec` is being parsed for `UI`, prior to any other command run. Thus the `start_date` must be populated from UI, and the `default` value for the spec does exactly this - the spec is being parsed by UI and pre-fiil the value from `default` value provided, like this:

```yaml
spec:
  type: Spec
  documentation_url: "https://docs.airbyte.com/integrations/sources/klaviyo"
  connection_specification:
    $schema: "http://json-schema.org/draft-07/schema#"
    title: "Klaviyo Spec"
    type: object
    properties:
      ....
      start_date:
        type: string
        title: "Start Date"
        description: "UTC date and time in the format 2017-01-25T00:00:00Z. Any data before this date will not be replicated. This field is optional - if not provided, all data will be replicated."
        pattern: "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$"
        # examples: ["2017-01-25T00:00:00Z"]
        default: "2012-01-01T00:00:00Z"
        format: date-time
        order: 1
        ....
```

So, for the local testing: both the `config` and the `spec` should specify the `start_date`, for the Cloud Testing or just UI - config can miss the `start_date` and it will be picked-up by the UI automatically, if the value isn't provided.

Reply:
Got it, thank you @bazarnov!